core/src/main/java/jenkins/security/CustomClassFilter.java

```diff
@@ -0,0 +1,177 @@
+ * Custom filters can be called frequently, and return values are uncached, so implementations should be fast.
+ * @see ClassFilterImpl
+ */
+@Restricted(NoExternalUse.class) // until a use case is identified for a _dynamic_ extension from some plugin
+public interface CustomClassFilter extends ExtensionPoint {
+
+    /**
+     * Determine whether a class should be permitted by {@link ClassFilter#isBlacklisted(Class)} of {@link ClassFilter#DEFAULT}.
+     * @param c the class
+     * @return true to permit it when it would normally be rejected (for example due to having a custom serialization method and being from a third-party library);
+     *         false to reject it when it would normally be permitted;
+     *         null to express no opinion (the default)
+     */
+    default @CheckForNull Boolean permits(Class<?> c) {
+        return null;
+    }
+
+    /**
+     * Determine whether a class should be permitted by {@link ClassFilter#isBlacklisted(String)} of {@link ClassFilter#DEFAULT}.
+     * @param name a class name
+     * @return true to permit it when it would normally be rejected (currently useless);
+     *         false to reject it when it would normally be permitted (currently due to {@link ClassFilter#STANDARD};
+     *         null to express no opinion (the default)
+     */
+    default @CheckForNull Boolean permits(String name) {
+        return null;
+    }
+
+    /**
+     * Standard filter which pays attention to a system property.
+     * To use, specify a system property {@code hudson.remoting.ClassFilter} containing a comma-separated list of {@link Class#getName} to whitelist.
+     * Entries may also be preceded by {@code !} to blacklist.
+     * Example: {@code -Dhudson.remoting.ClassFilter=com.google.common.collect.LinkedListMultimap,!com.acme.illadvised.YoloReflectionFactory$Handle}
+     */
+    @Restricted(DoNotUse.class)
+    @Extension
+    public class Static implements CustomClassFilter {
+
+        /**
+         * Map from {@link Class#getName} to true to permit, false to reject.
+         * Unmentioned classes are not treated specially.
+         * Intentionally {@code public} for possible mutation without restart by Groovy scripting.
+         */
+        public final Map<String, Boolean> overrides = new HashMap<>();
+
+        public Static() {
+            String entries = SystemProperties.getString("hudson.remoting.ClassFilter");
+            if (entries != null) {
+                for (String entry : entries.split(",")) {
+                    if (entry.startsWith("!")) {
+                        overrides.put(entry.substring(1), false);
+                    } else {
+                        overrides.put(entry, true);
+                    }
+                }
+                Logger.getLogger(Static.class.getName()).log(Level.FINE, "user-defined entries: {0}", overrides);
+            }
+        }
+
+        @Override
+        public Boolean permits(Class<?> c) {
```

Comment:
Would it make sense to support making decisions depending on a context? E.g. I may want to whitelist some classes for Remoting without allowing them to be saved to the disk via XStream. Hell, I want it for the most of the Remoting whitelist cases

Reply:
I am not aware of any case where such a contextual decision is needed. Either the class is potentially exploitable during deserialization, or it is not.